src/gui/fileactivitydialog.cpp

```diff
@@ -0,0 +1,231 @@
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+#include <QStringListModel>
+#include <QStyledItemDelegate>
+#include <QSpinBox>
+#include <QPainter>
+#include <QDebug>
+#include <QLoggingCategory>
+#include <QAbstractItemModel>
+#include <QPalette>
+#include <theme.h>
+
+#include "activityjob.h"
+#include "fileactivitydialog.h"
+#include "fileactivitymodel.h"
+#include "ui_fileactivitydialog.h"
+
+Q_LOGGING_CATEGORY(lcFileActivity, "nextcloud.gui.fileactivity", QtInfoMsg)
+
+namespace OCC {
+
+class FileActivityDelegate : public QStyledItemDelegate
+{
+    Q_OBJECT
+
+public:
+    FileActivityDelegate(QObject *parent = nullptr)
+        : QStyledItemDelegate(parent)
+    {
+        if (Theme::instance()->isHidpi()) {
+            _iconSize *= 2;
+        }
+    }
+
+    void paint(QPainter *painter, const QStyleOptionViewItem &option, const QModelIndex &index) const override
+    {
+        QStyleOptionViewItem opt(option);
+        initStyleOption(&opt, index);
+
+        painter->save();
+        painter->setClipping(true);
+        painter->setClipRect(opt.rect);
+        painter->setFont(opt.font);
+
+        drawBackground(opt, painter);
+        drawBottomLine(opt, index, painter);
+        const auto contentRect = opt.rect.adjusted(_margins.left(), _margins.top(), -_margins.right(), -_margins.bottom());
+        const auto item = qvariant_cast<FileActivity>(index.data());
+        drawMessageIcon(contentRect, item, painter);
+        const auto timestampRect = drawTimestamp(opt, contentRect, item, painter);
+        drawMessage(opt, timestampRect, item, painter);
+
+        painter->restore();
+    }
+
+    QSize sizeHint(const QStyleOptionViewItem &option, const QModelIndex &index) const override
+    {
+        QStyleOptionViewItem opt(option);
+        initStyleOption(&opt, index);
+
+        int textHeight = timestampBox(opt, index).height()
+            + _spacingVertical + messageBox(opt, index).height();
+        int iconHeight = _iconSize;
+        int h = textHeight > iconHeight ? textHeight : iconHeight;
+
+        return { opt.rect.width(), _margins.top() + h + _margins.bottom() };
+    }
+
+private:
+    QString timestampReadable(const QModelIndex &index) const
+    {
+        return timestampReadable(qvariant_cast<FileActivity>(index.data()).timestamp());
+    }
+
+    QString timestampReadable(const QDateTime &dateTime) const
+    {
+        return Utility::timeAgoInWords(dateTime.toLocalTime());
+    }
+
+    QRect timestampBox(const QStyleOptionViewItem &option, const FileActivity &fileActivity) const
+    {
+        QFont f(option.font);
```

Comment:
use auto here, I believe that calling `option.font` is explicit enough to know we are getting a `QFont` here